### PR TITLE
Fixed LBFGS giving NaN results

### LINKED ISSUE #5953: LBFGS always give nan results, why
I use the LBFGS alogorithm, and found that if maxiter is larger enough, i.e., maxiter >10, the optimizer always give nan results. Why?

### PR DESCRIPTION
**Pull request by abdrysdale**
If `ys` (inverse of rho_k) is very small, just set `ro` (rho) to be 1000. 

This is done to avoid very large gradient steps.

Solution similar to scipy:
https://github.com/scipy/scipy/blob/da64f8ca0ef2353b59994e7e37ecee4e67a9b1d3/scipy/optimize/_optimize.py#L1360-L1367

Fixes LBFGS NaN bug (https://github.com/pytorch/pytorch/issues/5953)

Fixes #5953

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
Hi, thanks for the change! Could you add a test to verifies this no longer runs into NaNs where it would have in the previous case?

I have a question for my own understanding: why do we move the history shift and H_diag outwards?

**Comment by abdrysdale** [comment]
> Hi, thanks for the change! Could you add a test to verifies this no longer runs into NaNs where it would have in the previous case?

Sure, I'll add a test ASAP.

> I have a question for my own understanding: why do we move the history shift and H_diag outwards?

This is simply to avoid writing it twice, once under the `if` statement and another time under the `else` statement.
The history steps do not need to be done inside the `if-else` statement so I just moved it out.

**Comment by abdrysdale** [comment]
I just noticed that the only tests for the LBFGS algorithm are checking if the type of the output is correct. Are there known input output combinations to use in LBFGS that can be implemented in tests?

I'm struggling to find a simple condition to test this NaN behaviour as I noticed it in my own work after about 1000 of iterations of an expensive evaluation function.

My apologies, my background is not in optimisation.

**Comment by janeyx99** [comment]
Hi! I see you've asked the experts on the issue thread and seem to have gotten a response already for the test.

I've been migrating tests from test/optim/test_optim.py to test/test_optim.py where you should see plenty more tests for LBFGS correctness + state_dictness. Feel free to add a standalone test for LBFGS in test/test_optim.py (a lot of the tests there tests ALL optim configs, and you probably only want to test the LBFGS configs).
